lib/pbench/agent/tool_meister_client.py

```diff
@@ -0,0 +1,227 @@
+            self.logger = logging.getLogger("tool-meister-client")
+        else:
+            self.logger = logger
+
+    def __enter__(self):
+        """On context entry, setup the connection with the Redis server.
+        """
+        if self.redis_server is None:
+            self.logger.debug("constructing Redis() object")
+            try:
+                self.redis_server = redis.Redis(
+                    host=self.redis_host, port=self.redis_port, db=0
+                )
+            except Exception as e:
+                self.logger.error(
+                    "Unable to connect to redis server, %s:%s: %s",
+                    self.redis_host,
+                    self.redis_port,
+                    e,
+                )
+                return 2
+            else:
+                self.logger.debug("constructed Redis() object")
+
+        if self.to_client_chan is None:
+            self.to_client_chan = RedisChannelSubscriber(
+                self.redis_server, self.to_client_channel
+            )
+        return self
+
+    def __exit__(self, *args):
+        """On context exit, just close down the to client channel object.
+        """
+        if self.to_client_chan is not None:
+            self.to_client_chan.close()
+
+    def publish(self, group, directory, action, args=None):
+        """publish - marshal a JSON document formed from the group, directory,
+        action, and args arguments.
+
+        Returns 0 on success, 1 on failure; logs are also written for any
+        errors encountered.
+        """
+        if action not in api_tm_allowed_actions:
+            return 1
+        # The published message contains four pieces of information:
+        #   {
+        #     "action": "< 'start' | 'stop' | 'send' | 'kill' >",
+        #     "group": "< the tool group name for the tools to operate on >",
+        #     "directory": "< the local directory path to store collected data >"
+        #     "args": "< arbitrary argument payload for a particular action >"
+        #   }
+        # The caller of tool-meister-client must be sure the directory argument
+        # is accessible by the Tool Data Sink instance.
+        self.logger.debug("publish %s on chan %s", action, self.from_client_channel)
+        msg = dict(action=action, group=group, directory=str(directory), args=args)
+        try:
+            num_present = self.redis_server.publish(
+                self.from_client_channel, json.dumps(msg)
+            )
+        except Exception:
+            self.logger.exception("Failed to publish client message")
+            return 1
+        else:
+            self.logger.debug("published %s", self.from_client_channel)
+            if num_present != 1:
+                self.logger.error(
+                    "Failed to publish to the TDS, encountered %d subscribers"
+                    " on the channel",
+                    num_present,
+                )
+                ret_val = 1
```

Comment:
Does a Redis pub/sub channel construct store the messages published in the channel if nobody is subscribed?

If it does, then we really have a problem with potential crud in the channel.  Before we start an instance the channels need to be flushed.

If it doesn't, then we'll never receive a response because the published message would have been dropped and we should just return immediately after posting the error.

Reply:
Tracked in https://github.com/distributed-system-analysis/pbench/issues/2160.